usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -206,41 +206,112 @@ typedef struct {
-                                    int retry_count;                     \
-                                    CK_RV rc2;                           \
-                                    if (((ep11_private_data_t *)         \
-                                            tokdata->private_data)->     \
-                                                         inconsistent) { \
-                                        (rc) = CKR_DEVICE_ERROR;         \
-                                        TRACE_ERROR("%s\n",              \
-                                             ock_err(ERR_DEVICE_ERROR)); \
-                                        break;                           \
-                                    }                                    \
-                                    ep11_target_info_t* target_info =    \
-                                             get_target_info((tokdata)); \
-                                    if (target_info == NULL) {           \
-                                        (rc) = CKR_FUNCTION_FAILED;      \
-                                        break;                           \
-                                    }                                    \
-                                    for(retry_count = 0;                 \
-                                        target_info != NULL &&           \
-                                        retry_count < MAX_RETRY_COUNT;   \
-                                        retry_count ++) {
-
-#define RETRY_END(rc, tokdata, session)  if ((rc) != CKR_SESSION_CLOSED) \
-                                             break;                      \
-                                         rc2 = ep11tok_relogin_session(  \
-                                                  (tokdata), (session)); \
-                                         if (rc2 != CKR_OK) {            \
-                                             (rc) = rc2;                 \
-                                             break;                      \
-                                         }                               \
-                                    }                                    \
-                                    put_target_info((tokdata),           \
-                                                    target_info);        \
-                                } while (0);
+/*
+ * Macros to enclose EP11 library calls involving session bound blobs.
+ * If the EP11 token is in an inconsistent state, fail with CKR_DEVICE_ERROR.
+ * Obtain a target_info to be used with the EP11 library call.
+ * If in single-APQN mode, and that APQN went offline, select another APQN and
+ * retry the library call.
+ * In case of EP11 library function failed with CKR_SESSION_CLOSED, relogin
+ * all APQNs and retry the library call.
+ */
+#define RETRY_SESSION_SINGLE_APQN_START(rc, tokdata)                     \
+                do {                                                     \
```

Comment:
Why so much indention in the macro? It affects readability ...

Reply:
well, mostly because these macros have already been that way....